### PR TITLE
Fix ASAN failure in TestSingleThreadStepTimeout.py

### PR DESCRIPTION
**Pull request by jeffreytan81**
This PR fixes the ASAN failure in https://github.com/llvm/llvm-project/pull/90930. 

The original PR made the assumption that parent `ThreadPlanStepOverRange`'s lifetime will always be longer than `ThreadPlanSingleThreadTimeout` leaf plan so it passes the `m_timeout_info` as reference to it. 
From the ASAN failure, it seems that this assumption may not be true (likely the thread stack is holding a strong reference to the leaf plan). 

This PR fixes this lifetime issue by using shared pointer instead of passing by reference.

### CODE REVIEW

**Review by jimingham — APPROVED** [review]
This looks fine.  I'd add a 

using  TimeoutInfoSP = std::shared_ptr<...>

and refer to it that way.  The construct appears enough times to warrant that; it's a lot easier to read and we use the SP suffix consistently so it will be understood.  But that's a matter of taste.

**Comment by jeffreytan81** [comment]
I am still waiting for all bots to green before landing it. Let me know if I should skip and land immediately to unblock anyone.

**Pull request merged**
